From efb2bc578658427d578acf0024ae706bc9238a53 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Mon, 8 Jan 2024 15:48:44 +0100
Subject: [PATCH 078/120] remoteproc: stm32_rproc: save the phys resource table
 address in backup reg

The resource table address can be retrieved from the ELF file or obtained
from OP-TEE. The OP-TEE interface provides the physical address.
To ensure coherence, the device address obtained from the ELF file
should be converted to a physical address before being saved to the
backup register.

Converting the physical address provides by OP-TEE to the device address,
is not simple as the driver can not have the information in the DT.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git 838791a70a66032dde2b645fbba206ee6df88bc2 remoteproc: stm32_rproc: save the phys resource table address in backup reg]
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Change-Id: I03ea2bd30bc55748ed74beaba9f1f03a48b8e92e
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/350379
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/359778
Tested-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Domain-Review: Patrice CHOTARD <patrice.chotard@foss.st.com>
---
 drivers/remoteproc/stm32_copro.c | 58 ++++++++++++++++++++------------
 1 file changed, 37 insertions(+), 21 deletions(-)

diff --git a/drivers/remoteproc/stm32_copro.c b/drivers/remoteproc/stm32_copro.c
index bb6b3ec737a..20525eb0b27 100644
--- a/drivers/remoteproc/stm32_copro.c
+++ b/drivers/remoteproc/stm32_copro.c
@@ -112,6 +112,32 @@ static int stm32_copro_remove(struct udevice *dev)
 	return 0;
 }
 
+/**
+ * stm32_copro_device_to_phys() - Convert device address to physical address
+ * @dev:	corresponding STM32 remote processor device
+ * @da:		device address
+ * @size:	Size of the memory region @da is pointing to
+ * Return: converted physical address
+ */
+static phys_addr_t stm32_copro_device_to_phys(struct udevice *dev, ulong da,
+					      ulong size)
+{
+	fdt32_t in_addr = cpu_to_be32(da), end_addr;
+	phys_addr_t paddr;
+
+	paddr = dev_translate_dma_address(dev, &in_addr);
+	if (paddr == OF_BAD_ADDR) {
+		dev_err(dev, "Unable to convert address %ld\n", da);
+		return 0;
+	}
+	end_addr = cpu_to_be32(da + size - 1);
+	if (dev_translate_dma_address(dev, &end_addr) == OF_BAD_ADDR) {
+		dev_err(dev, "Unable to convert address %ld\n", da + size - 1);
+		return 0;
+	}
+
+	return paddr;
+}
 /**
  * stm32_copro_device_to_virt() - Convert device address to virtual address
  * @dev:	corresponding STM32 remote processor device
@@ -122,25 +148,11 @@ static int stm32_copro_remove(struct udevice *dev)
 static void *stm32_copro_device_to_virt(struct udevice *dev, ulong da,
 					ulong size)
 {
-	fdt32_t in_addr = cpu_to_be32(da), end_addr;
-	unsigned int proc_id = (u32)dev_get_driver_data(dev);
 	phys_addr_t paddr;
 
-	if (proc_id == STM32MP15_M4_FW_ID) {
-		paddr = dev_translate_dma_address(dev, &in_addr);
-		if (paddr == OF_BAD_ADDR) {
-			dev_err(dev, "Unable to convert address %ld\n", da);
-			return NULL;
-		}
-		end_addr = cpu_to_be32(da + size - 1);
-		if (dev_translate_dma_address(dev, &end_addr) == OF_BAD_ADDR) {
-			dev_err(dev, "Unable to convert address %ld\n", da + size - 1);
-			return NULL;
-		}
-	} else {
-		/* No translation */
-		paddr = (phys_addr_t)da;
-	}
+	paddr = stm32_copro_device_to_phys(dev, da, size);
+	if (!paddr)
+		return NULL;
 
 	return phys_to_virt(paddr);
 }
@@ -157,6 +169,8 @@ static int stm32_copro_load(struct udevice *dev, ulong addr, ulong size)
 	struct stm32_copro_privdata *priv = dev_get_priv(dev);
 	struct rproc_optee *trproc = &priv->trproc;
 	ulong rsc_table_size = 0;
+	ulong rsc_table_addr = 0;
+	phys_addr_t paddr;
 	int ret;
 
 	if (trproc->tee)
@@ -174,17 +188,19 @@ static int stm32_copro_load(struct udevice *dev, ulong addr, ulong size)
 		return ret;
 	}
 
-	ret = rproc_elf32_load_rsc_table(dev, addr, size, &priv->rsc_table_addr,
+	priv->rsc_table_addr = 0;
+	ret = rproc_elf32_load_rsc_table(dev, addr, size, &rsc_table_addr,
 					 &rsc_table_size);
 	if (ret) {
 		if (ret != -ENODATA)
 			return ret;
-
 		dev_dbg(dev, "No resource table for this firmware\n");
-		priv->rsc_table_addr = 0;
 	}
-	priv->rsc_table_size = rsc_table_size;
 
+	paddr = stm32_copro_device_to_phys(dev, rsc_table_addr, rsc_table_size);
+
+	priv->rsc_table_addr = (ulong)paddr;
+	priv->rsc_table_size = rsc_table_size;
 
 	return rproc_elf32_load_image(dev, addr, size);
 }
-- 
2.47.2

